Add UptimeKuma Service component

adyanth opened the pull request:
## Description
Adds Uptime Kuma service component.

Fixes #407

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] I've read & comply with the [contributing guidelines](https://github.com/bastienwirtz/homer/blob/main/CONTRIBUTING.md)
- [X] I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers. 
- [X] I have made corresponding changes to the documentation (README.md).
- [X] I've checked my modifications for any breaking changes, especially in the `config.yml` file

adyanth commented:
I can add this too. The partial status shows when the last heartbeat for any of the listed services is down.
https://github.com/louislam/uptime-kuma/blob/38e6e846bfc1598505233b7093fc946ef4320a64/src/pages/StatusPage.vue#L351

adyanth commented:
With the new release of Uptime Kuma, multiple status page support has been added and the APIs have changed. Will need to add a new field for the slug and fix the URLs.

bastienwirtz commented on src/components/services/UptimeKuma.vue:
```diff
@@ -0,0 +1,110 @@
+  },
+  methods: {
+    fetchStatus: async function () {
+      const incident = await this.fetch("/api/status-page/incident").catch(
+        (e) => console.log(e)
+      );
+      this.incident = incident;
+
+      const heartbeat = await this.fetch("/api/status-page/heartbeat").catch(
+        (e) => console.log(e)
+      );
+      this.heartbeat = heartbeat;
```
Could you use a `.then()` instead of `await` ? That way both requests can run in parallel instead of one after the other.

bastienwirtz reviewed (commented):
Looking good so far! Thanks for your work @adyanth.

adyanth commented:
Converted to the new API format from Uptime Kuma and added the tri-state logic. 
* If an Incident is found, it will be "bad" with the incident title
* If there are no incidents but there is a service down it will be "warn" with "Partially Degraded Service"
* If no incidents and all services are up it will be a "good" and "All Systems Operational"

Roundaround commented on docs/customservices.md:
```diff
@@ -10,7 +10,6 @@ within Homer.
 
 If you experiencing any issue, please have a look to the [troubleshooting](troubleshooting.md) page.
 
-
```
Small nit: undo unrelated change. Here and lines 82 & 135

adyanth replied:
That was my markdown linter, sorry :)

Roundaround commented on docs/customservices.md:
```diff
@@ -141,3 +140,20 @@ See https://docs.portainer.io/v/ce-2.11/user/account-settings#access-tokens
+
+Using the Uptime Kuma service you can display info about your instance uptime right on your Homer dashboard.
+
+The following configuration is available for the UptimeKuma service. Needs v1.13.1 or later because of the change in APIs due to [multiple status pages support](https://github.com/louislam/uptime-kuma/releases/tag/1.13.1).
```
Does the whole custom service require the 1.13.1 API or just the slug/multi status page support?

If the whole thing requires it, I'd recommend rephrasing this line to just simply be "Requires Uptime Kuma version 1.13.1 or later." and scrapping the mention of the config or the link out to their repo.

If only the slug feature requires it, clarify as such here and as a comment in the example yaml file

adyanth replied:
The whole service needs it. 1.13.1 changed the API structure along with introducing multiple status pages. So the component can support either the old API format or the new one, not both. So went with the new format.

Roundaround commented on src/components/services/UptimeKuma.vue:
```diff
@@ -0,0 +1,153 @@
+      let hasUp = false;
+      for (let id in this.lastHeartBeatList) {
+        let beat = this.lastHeartBeatList[id];
+        console.log(beat.status);
```
```suggestion
```

Remove console.log

adyanth replied:
Haha, my bad :) Troubleshooting the API response there

Roundaround commented on docs/customservices.md:
```diff
@@ -141,3 +140,20 @@ See https://docs.portainer.io/v/ce-2.11/user/account-settings#access-tokens
+Using the Uptime Kuma service you can display info about your instance uptime right on your Homer dashboard.
+
+The following configuration is available for the UptimeKuma service. Needs v1.13.1 or later because of the change in APIs due to [multiple status pages support](https://github.com/louislam/uptime-kuma/releases/tag/1.13.1).
+
+As usual, needs CORS to be enabled from the homer domain to uptime kuma.
```
```suggestion
```

Another small nit: we don't mention CORS anywhere else on this page, and in general, we're looking to address the CORS issue with better documentation and potentially a first-party CORS proxy soon (TM)

adyanth replied:
Ooo that's exciting! Looking forward to that.

Roundaround commented on src/components/services/UptimeKuma.vue:
```diff
@@ -0,0 +1,153 @@
+  methods: {
+    fetchStatus: function () {
+      this.fetch(`/api/status-page/${this.dashboard}`)
+        .catch((e) => console.log(e))
```
```suggestion
        .catch((e) => console.error(e))
```

Roundaround commented on src/components/services/UptimeKuma.vue:
```diff
@@ -0,0 +1,153 @@
+        .then((resp) => (this.incident = resp));
+
+      this.fetch(`/api/status-page/heartbeat/${this.dashboard}`)
+        .catch((e) => console.log(e))
```
```suggestion
        .catch((e) => console.error(e))
```

Roundaround reviewed (changes requested):
Looking good! Just a few suggestions

Roundaround commented on src/components/services/UptimeKuma.vue:
```diff
@@ -0,0 +1,153 @@
+  },
+  methods: {
+    fetchStatus: function () {
+      this.fetch(`/api/status-page/${this.dashboard}`)
```
Sometimes the Uptime Kuma API will return 304 if nothing has changed. Since we don't store any state or have a local cache, we need to bust the cache with our requests. Easiest way to do that is with a url param. Add here and below

```suggestion
      this.fetch(`/api/status-page/${this.dashboard}?cachebust=${Date.now()}`)
```

adyanth replied:
Does Uptime Kuma do that? It might be a proxy in between since I did not see that happening. Either way, the cache bust is a good idea, thanks!

Roundaround commented on src/components/services/UptimeKuma.vue:
```diff
@@ -0,0 +1,153 @@
+        .catch((e) => console.log(e))
+        .then((resp) => (this.incident = resp));
+
+      this.fetch(`/api/status-page/heartbeat/${this.dashboard}`)
```
```suggestion
      this.fetch(`/api/status-page/heartbeat/${this.dashboard}?cachebust=${Date.now()}`)
```

Roundaround reviewed (changes requested):
Sorry, just tried the integration myself locally and ran into some issues with Uptime Kuma returning 304s. Left a couple more comments!

adyanth commented:
Let me know if this looks good @Roundaround , thanks for those suggestions!

Roundaround commented:
Looks good, thanks for the fast response!
